--- libgmail-0.0.8/libgmail.py.orig	2005-07-27 19:25:06.000000000 +0200
+++ libgmail-0.0.8/libgmail.py	2005-07-27 19:20:09.000000000 +0200
@@ -274,7 +274,7 @@
 
         # TODO: Tidy this up?
         # This requests the page that provides the required "GV" cookie.
-        RE_PAGE_REDIRECT = 'top\.location\W=\W"CheckCookie\?continue=([^"]+)'
+        RE_PAGE_REDIRECT = 'top\.location\W=.*CheckCookie\?continue=([^"]+)'
         # TODO: Catch more failure exceptions here...?
         try:
             redirectURL = urllib.unquote(re.search(RE_PAGE_REDIRECT,
